knitr/math-guide/math_guide.Rmd

```diff
@@ -0,0 +1,643 @@
+it new meta-programs should be added.
+
+## Prim (fully templated)
+
+### Exceptions
+
+Math functions are expected to validate inputs and throw exceptions.
+It should be impossible to cause a segfault by calling a
+user-facing Math function or trigger an exception that is not handled
+by Stan itself. This means that argument sizes should be checked to be
+compatible. For instance, a matrix multiply function would need to check
+that the first matrix can in fact be multiplied by the second, if this
+is not so it should throw.
+
+If a scalar argument should be positive, an exception should be thrown
+if it is not. If a matrix argument should be positive definite, the
+function should check this before operating on the matrix. This can
+lead to significant overhead in the checks, but the current library
+policy is to get informative errors back to the user, which means doing
+these checks.
+
+Error checks should be done with the functions in
+[stan/math/prim/err](https://github.com/stan-dev/math/tree/develop/stan/math/prim/err)
+which throw consistently formatted errors.
+
+### Reference types
+
+Any time a function takes a vector or matrix type, it needs to be able
+to handle an expression. Some expressions can be expensive to evaluate, so
+when they are evaluated, they should be saved. This includes any arithmetic
+operation, a matrix multiply, or a sum, or something else. Some expressions
+are cheap to evaluate though, any of the Eigen views quality here (transpose,
+block access, segment, etc). In this case, evaluating the expression is not
+necessary -- it would only lead to another allocate and copy.
+
+The Math function `to_ref` is a solution for this. For a vector
+or matrix variable `x`, `to_ref(x)` returns an Eigen reference to the
+input variable `x`. If `x` was an expensive expression, it will be evaluated.
+If it was a cheap expression, the reference type won't evaluate. If it
+isn't an Eigen type, the `to_ref` just forwards.
+
+```cpp
+template <typename T,
+          require_eigen_t<T>* = nullptr>
+auto myfunc(const T& x) {
+  const auto& x_ref = to_ref(x);
+}
+```
+
+### Holders
+
+If a function returns an Eigen expression, it may be necessary to use
+the `make_holder` utility.
+
+Returning expressions is tricky because the expression may have a longer
+lifetime than the objects it operates on. For instance, returning an
+expression on a local variable leads to this situation:
+
+```cpp
+template <typename T>
+auto add_zero(const T& a) {
+  Eigen::VectorXd b = Eigen::VectorXd(a.size());
+  return a + b;
+}
+```
+
+Because the return type of this function is `auto`, it is possible
+that it returns an Eigen expression.
+
+The following code will segfault on the construction of `c` because
+by this time the temporary `b` will have been destructed.
+
+```
+Eigen::VectorXd a(5);
+Eigen::VectorXd c = add_zero(a);
+```
+
+`make_holder` can be used to extend the scope of `b` so that it does
+not immediately get destructed.
```

Comment:
```suggestion
`make_holder` can be used to extend the lifetime of `b` so that it matches the expression it is used in and does
not immediately get destructed.
```